storagestate/storagestate.go

```diff
@@ -0,0 +1,730 @@
+}
+
+// Setup setup storagestate instance.
+func (storageState *StorageState) Setup(
+	params SetupParams,
+) (storagePath string, statePath string, err error) {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	instancePath := fmt.Sprintf("%d_%s_%s", params.Instance, params.SubjectID, params.ServiceID)
+
+	log.WithFields(log.Fields{
+		"instance":     params.Instance,
+		"serviceID":    params.ServiceID,
+		"subjectID":    params.SubjectID,
+		"storageQuota": params.StorageQuota,
+		"stateQuota":   params.StateQuota,
+	}).Debug("Setup storage and state")
+
+	storageStateInfo, err := storageState.storage.GetStorageStateInfo(instancePath)
+	if err != nil {
+		if !errors.Is(err, ErrNotExist) {
+			return "", "", aoserrors.Wrap(err)
+		}
+
+		storageStateInfo = StorageStateInstanceInfo{
+			InstancePath:  instancePath,
+			InstanceIdent: params.InstanceIdent,
+		}
+
+		if err = storageState.storage.AddStorageStateInfo(storageStateInfo); err != nil {
+			return "", "", aoserrors.Wrap(err)
+		}
+	}
+
+	if storagePath, err = storageState.prepareStorage(instancePath, params); err != nil {
+		return "", "", aoserrors.Wrap(err)
+	}
+
+	if err := storageState.stopStateWatching(params.InstanceIdent); err != nil {
+		return "", "", aoserrors.Wrap(err)
+	}
+
+	if storageStateInfo.StateQuota != params.StateQuota || storageStateInfo.StorageQuota != params.StorageQuota {
+		if err = storageState.setQuotasFS(params); err != nil {
+			return "", "", aoserrors.Wrap(err)
+		}
+
+		if err = storageState.storage.SetStorageStateQuotas(
+			instancePath, params.StorageQuota, params.StateQuota); err != nil {
+			return "", "", aoserrors.Wrap(err)
+		}
+	}
+
+	if params.StateQuota != 0 {
+		defer func() {
+			if err != nil {
+				if err := storageState.stopStateWatching(params.InstanceIdent); err != nil {
+					log.Warnf("Can't stop watching state file: %v", err)
+				}
+			}
+		}()
+
+		if statePath, err = storageState.prepareState(
+			instancePath, params, storageStateInfo.StateChecksum); err != nil {
+			return "", "", aoserrors.Wrap(err)
+		}
+	} else {
+		if err := os.RemoveAll(
+			path.Join(storageState.stateDir, fmt.Sprintf(stateFileFormat, instancePath))); err != nil {
+			return "", "", aoserrors.Wrap(err)
+		}
+	}
+
+	return storagePath, statePath, nil
+}
+
+// Cleanup clean storagestate instance.
+func (storageState *StorageState) Cleanup(instanceIdent aostypes.InstanceIdent) error {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	log.WithFields(log.Fields{
+		"instance":  instanceIdent.Instance,
+		"serviceID": instanceIdent.ServiceID,
+		"subjectID": instanceIdent.SubjectID,
+	}).Debug("Clean storage and state")
+
+	if err := storageState.stopStateWatching(instanceIdent); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+// Remove remove storagestate dirs.
+func (storageState *StorageState) Remove(serviceID string) error {
+	stateStorageInfos, err := storageState.storage.GetAllStorageStateInfo()
```

Comment:
this function should be guarded with mutex as well.

Reply:
Mutex is locked in the `Clean `function